### PR TITLE
[Search experience] Update notebooks and docs with ELSER v2

### PR DESCRIPTION
**Pull request by saarikabhasi**
This PR updates :
1. [ELSER](https://github.com/elastic/elasticsearch-labs/blob/main/notebooks/search/03-ELSER.ipynb) notebook to  use ELSER v2 `.elser_model_2` and use `sparse_vector`
2. Updates all the usage of ELSER model `.elser_model_1` to use `.elser_model_2`
3. Update [relevance-workbench/README.md](https://github.com/elastic/elasticsearch-labs/blob/main/example-apps/relevance-workbench/README.md) with latest UI
4. Creates a new notebook under the new folder :` /notebooks/model-upgrades/` with steps on [how to upgrade your index to an ELSER model](https://github.com/saarikabhasi/elasticsearch-labs/blob/saarikabhasi/update-notebooks-doc-to-ELSERv2/notebooks/model-upgrades/upgrading-index-to-use-elser.ipynb) 
5. Update Readme with new models upgrading file path
6. Tested Upgrading and ELSER notebook with the [testing framework](https://github.com/elastic/elasticsearch-labs/tree/main/test) and update [Makefile](https://github.com/elastic/elasticsearch-labs/blob/main/Makefile) to include [ELSER](https://github.com/elastic/elasticsearch-labs/blob/main/notebooks/search/03-ELSER.ipynb) notebook

### CODE REVIEW

File: notebooks/langchain/langchain-vector-store-using-elser.ipynb [hunk 1]
```diff
@@ -4,11 +4,11 @@
    "cell_type": "markdown",
```
Comment (joemcelroy) [inline notebooks/langchain/langchain-vector-store-using-elser.ipynb:4]: you also need to specify the name of the model to use it from the default elser v1
```
    strategy=ElasticsearchStore.SparseVectorRetrievalStrategy(model_id=".elser_model_2")
```
double check the model id, im guessing here.

File: notebooks/search/03-ELSER.ipynb [hunk 2]
```diff
@@ -12,7 +12,7 @@
     "[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/elastic/elasticsearch-labs/blob/main/notebooks/search/03-ELSER.ipynb)\n",
     "\n",
     "\n",
-    "Learn how to use the [ELSER](https://www.elastic.co/guide/en/machine-learning/current/ml-nlp-elser.html) for text expansion-powered semantic search."
+    "Learn how to use the [ELSER v2](https://www.elastic.co/guide/en/machine-learning/master/ml-nlp-elser.html) for text expansion-powered semantic search."
```
Comment (joemcelroy) [inline notebooks/search/03-ELSER.ipynb:15]: keep it as current as when 8.11 is released, this current page link will be updated.

File: notebooks/langchain/langchain-vector-store-using-elser.ipynb (hunk 1 above)
Comment (saarikabhasi) [reply notebooks/langchain/langchain-vector-store-using-elser.ipynb:4]: Good catch. Will update 👍

File: notebooks/langchain/langchain-vector-store-using-elser.ipynb [hunk 3]
```diff
@@ -86,11 +86,11 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 29,
+   "execution_count": 4,
    "metadata": {},
    "outputs": [],
    "source": [
-    "url = \"https://raw.githubusercontent.com/elastic/elasticsearch-labs/main/example-apps/workplace-search/example-data/data.json\"\n",
+    "url = \"https://raw.githubusercontent.com/elastic/elasticsearch-labs/main/example-apps/workplace-search/data/data.json\"\n",
```
Comment (saarikabhasi) [inline notebooks/langchain/langchain-vector-store-using-elser.ipynb:93]: This change is not related to this PR. But thought of adding it as the notebook was using wrong path.

File: notebooks/langchain/langchain-vector-store-using-elser.ipynb [hunk 4]
```diff
@@ -138,18 +138,18 @@
...
    "cell_type": "code",
-   "execution_count": 45,
+   "execution_count": 7,
    "metadata": {},
    "outputs": [],
    "source": [
     "documents = vector_store.from_documents(\n",
     "    docs, es_cloud_id=CLOUD_ID, es_user=CLOUD_USERNAME, es_password=CLOUD_PASSWORD, index_name=\"workplace_index\",\n",
-    "    strategy=ElasticsearchStore.SparseVectorRetrievalStrategy()\n",
+    "    strategy=ElasticsearchStore.SparseVectorRetrievalStrategy(\".elser_model_2\")\n",
```
Comment (joemcelroy) [inline notebooks/langchain/langchain-vector-store-using-elser.ipynb:152]: i think this needs to be a named parameter here. (model_id=".elser_model_2")

File: README.md [hunk 5]
```diff
@@ -41,6 +41,8 @@ The [`notebooks`](notebooks/README.md) folder contains a range of executable Pyt
 - [`02-hybrid-search.ipynb`](./notebooks/search/02-hybrid-search.ipynb)
 - [`03-ELSER.ipynb`](./notebooks/search/03-ELSER.ipynb)
 - [`04-multilingual.ipynb`](./notebooks/search/04-multilingual.ipynb)
+- [`05-query-rules.ipynb`](./notebooks/search/05-query-rules.ipynb)
```
Comment (saarikabhasi) [inline README.md:44]: Added Query rules file as it was missing

**Comment by saarikabhasi** [comment]
🔔  Merge this PR when 8.11 is released

File: example-apps/relevance-workbench/README.md [hunk 6]
```diff
@@ -10,7 +10,7 @@ You can also try with your own data by forking this repo and plugging the applic
 
 ## Pre-requisites
 
-To run this demo successfully, you will need an Elasticsearch deployment (> 8.8) with the ELSER model deployed. The easiest way for this is to use Elastic Cloud as described in the next part but you can also deploy Elasticsearch locally.
+To run this demo successfully, you will need an Elasticsearch deployment (> 8.11) with the ELSER model deployed. The easiest way for this is to use Elastic Cloud as described in the next part but you can also deploy Elasticsearch locally.
```
Comment (carlosdelest) [inline example-apps/relevance-workbench/README.md:13]: Should we also point out that >8.8 has ELSERv1, and >8.11 has ELSERv2? We might want users that have not been able to upgrade yet to be able to use this notebook

File: notebooks/search/03-ELSER.ipynb [hunk 7]
```diff
@@ -57,16 +57,17 @@
    },
    "source": [
     "# Setup ELSER\n",
-    "To use ELSER, you must have the [appropriate subscription]() level\n",
+    "\n",
+    "From 8.11, Elastic supports  `.elser_model_2` model which offers an improved retrieval accuracy. To use ELSER, you must have the [appropriate subscription]() level\n",
     "for semantic search or the trial period activated.\n",
     "\n",
-    "Follow these [instructions](https://www.elastic.co/guide/en/machine-learning/8.8/ml-nlp-elser.html#trained-model) to download and deploy ELSER in the Kibana UI or using the Dev Tools **Console**.\n",
+    "Follow these [instructions](https://www.elastic.co/guide/en/machine-learning/current/ml-nlp-elser.html#trained-model) to download and deploy ELSER  `.elser_model_2` in the Kibana UI or using the Dev Tools **Console**.\n",
```
Comment (carlosdelest) [inline notebooks/search/03-ELSER.ipynb:64]: As stated above, I think we could add support for ELSER v1 as a separate section, or as an additional script, so v1 users can use the notebook

File: notebooks/search/03-ELSER.ipynb [hunk 8]
```diff
@@ -75,7 +76,13 @@
     "``` -->\n",
     "<!-- # deploy model\n",
     "```json\n",
-    "POST _ml/trained_models/.elser_model_1/deployment/_start?deployment_id=for_search -->"
+    "POST _ml/trained_models/.elser_model_2/deployment/_start?deployment_id=for_search -->\n",
+    "\n",
+    "## Upgrading existing index to `.elser_model_2` ELSER model\n",
+    "\n",
+    "If have already had set up an index with ELSER model `.elser_model_1`, You can easily upgrade to  `.elser_model_2`  model by [Reindexing with ingest pipeline](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-reindex.html#reindex-with-an-ingest-pipeline) referencing ELSER `.elser_model_2` and starting querying your data with ELSER right away. \n",
```
Comment (carlosdelest) [inline notebooks/search/03-ELSER.ipynb:83]:
```suggestion
    "If have already had set up an index with ELSER model `.elser_model_1`, you can easily upgrade to  `.elser_model_2`  model by [Reindexing with ingest pipeline](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-reindex.html#reindex-with-an-ingest-pipeline) referencing ELSER `.elser_model_2` and starting querying your data with ELSER right away. \n",
```

Comment (carlosdelest) [inline notebooks/search/03-ELSER.ipynb:83]: We can also reference [update by query](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update-by-query.html) as a way to update indices in place. We might want to do something like:

```suggestion
    "If have already had set up an index with ELSER model `.elser_model_1`, You can easily upgrade to  `.elser_model_2`  model by\n:"
    " -  [Reindexing with ingest pipeline](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-reindex.html#reindex-with-an-ingest-pipeline) to ingest into a separate index\n"
    " - [[Update by query](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update-by-query.html)] using an ingest pipeline to update the index in place\n"
    "In both cases, you will need to reference ELSER `.elser_model_2`.\n"
     "After reindexing or updating, you can start querying your data with ELSER right away. \n",
```

File: example-apps/relevance-workbench/README.md (hunk 6 above)
Comment (leemthompo) [reply example-apps/relevance-workbench/README.md:13]: Maybe we should just have _separate_ notebooks for simplicity? v1 and v2, with a big info banner at top of each explaining version availability etc.?

Comment (carlosdelest) [reply example-apps/relevance-workbench/README.md:13]: I like that - we could have a "if you're using version < 8.11, then go here"

File: notebooks/search/03-ELSER.ipynb [hunk 9]
```diff
@@ -305,7 +307,7 @@
     "Let's note a few important parameters from that API call:\n",
     "\n",
     "- `inference`: A processor that performs inference using a machine learning model.\n",
-    "- `model_id`: Specifies the ID of the machine learning model to be used. In this example, the model ID is set to `.elser_model_1`.\n",
+    "- `model_id`: Specifies the ID of the machine learning model to be used. In this example, the model ID is set to `.elser_model_2`.\n",
     "- `target_field`: Defines the field where the inference result will be stored. Here, it is set to `ml`.\n",
```
Comment (brianjolly) [inline notebooks/search/03-ELSER.ipynb:311]: Hi @saarikabhasi! Should `target_field` here be updated to `output_field` to match [the JSON above](https://github.com/elastic/elasticsearch-labs/pull/72/files#diff-cefcd1253dbd13c7f622213b0e43e549c91cc4e4b4634996cb227e75f52ea5f8R291),

and maybe remove `text_expansion`, since `inference_config` isn't in the call anymore?

Comment (saarikabhasi) [reply notebooks/search/03-ELSER.ipynb:311]: Thanks @brianjolly. I might have missed to update the parameter description. Will update the PR to include the suggested change. Thanks again!

File: notebooks/search/03-ELSER.ipynb (hunk 8 above)
Comment (saarikabhasi) [reply notebooks/search/03-ELSER.ipynb]: Thanks @carlosdelest I added a note in [upgrading-index-to-use-elser notebook](https://github.com/elastic/elasticsearch-labs/blob/bc36bfa5d1ce56dc1e7412bc8a91db5ba1c36751/notebooks/model-upgrades/upgrading-index-to-use-elser.ipynb)
Commit bc36bfa5d1ce56dc1e7412bc8a91db5ba1c36751

Comment (saarikabhasi) [reply notebooks/search/03-ELSER.ipynb]: Simplified the sentence with 

> If you have set up an index with ELSER model .elser_model_1, and would like to upgrade to ELSER v2 model - .elser_model_2, Please follow instructions from the notebook on [how to upgrade an index to use elser model](https://github.com/elastic/elasticsearch-labs/blob/bc36bfa5d1ce56dc1e7412bc8a91db5ba1c36751/notebooks/model-upgrades/upgrading-index-to-use-elser.ipynb)

Commit https://github.com/elastic/elasticsearch-labs/commit/bc36bfa5d1ce56dc1e7412bc8a91db5ba1c36751

File: example-apps/relevance-workbench/README.md (hunk 6 above)
Comment (saarikabhasi) [reply example-apps/relevance-workbench/README.md:13]: Based on the outcome of discussion in [thread](https://elastic.slack.com/archives/C05E3NSDLS2/p1697556832684659?thread_ts=1696523296.981769&cid=C05E3NSDLS2), we are not adding/updating notebook(s) to support ELSER v1 model.

File: notebooks/search/03-ELSER.ipynb (hunk 7 above)
Comment (saarikabhasi) [reply notebooks/search/03-ELSER.ipynb]: Same as above. 

> Based on the outcome of discussion in [thread](https://elastic.slack.com/archives/C05E3NSDLS2/p1697556832684659?thread_ts=1696523296.981769&cid=C05E3NSDLS2), we will be focussing one one notebook for ELSER v2 model.

**Comment by joemcelroy** [comment]
reviewed last changes, look great! Just one suggestion, rather than deleting the elser model, could we check to see if it exists and skip download? 

```
# delete model if already downloaded and deployed
try:
  client.ml.delete_trained_model(model_id=".elser_model_2",force=True)
  print("Model deleted successfully, We will proceed with creating one")
except exceptions.NotFoundError:
  print("Model doesn't exist, but We will proceed with creating one")

# Creates the ELSER model configuration. Automatically downloads the model if it doesn't exist. 
client.ml.put_trained_model(
    model_id=".elser_model_2",
    input={
      "field_names": ["text_field"]
    }
  )
```

**Comment by saarikabhasi** [comment]
> reviewed last changes, look great! Just one suggestion, rather than deleting the elser model, could we check to see if it exists and skip download?

We could do that but [GET trained model API](https://www.elastic.co/guide/en/elasticsearch/reference/current/get-trained-models.html) returns 404 if the model doesn't exist. 

[Start trained model deployment API](https://www.elastic.co/guide/en/elasticsearch/reference/current/start-trained-model-deployment.html) will also fail if model already deployed or doesn't exist. To handle this, we would have to save status of this GET and check this flag in download model and start deployment cell. 

When the model is recreated, if there are any indices that use this model, text_expansion query works. Tested in cloud.

**Comment by miguelgrinberg** [comment]
@saarikabhasi You can also add this notebook to notebooks/search/Makefile so that it runs when you run `make` from the top.

**Comment by saarikabhasi** [comment]
> @saarikabhasi You can also add this notebook to notebooks/search/Makefile so that it runs when you run `make` from the top.

updated in a3a7cc6c0d2b5bc0b0b9305c826b4471be69c0f2

**Review by miguelgrinberg — APPROVED** [review]
